qutip/solver/countstat.py

```diff
@@ -0,0 +1,225 @@
+"""
+This module contains functions for calculating current and current noise using
+the counting statistics formalism.
+"""
+__all__ = ['countstat_current', 'countstat_current_noise']
+
+import numpy as np
+import scipy.sparse as sp
+
+from itertools import product
+from ..core import (
+    sprepost, spre, operator_to_vector, identity, tensor,
+)
+from ..core import data as _data
+from qutip import pseudo_inverse, steadystate
+from ..settings import settings
+
+# Load MKL spsolve if avaiable
+if settings.has_mkl:
+    from qutip._mkl.spsolve import mkl_spsolve
+
+
+def countstat_current(L, c_ops=None, rhoss=None, J_ops=None):
+    """
+    Calculate the current corresponding a system Liouvillian `L` and a list of
+    current collapse operators `c_ops` or current superoperators `J_ops`
+    (either must be specified). Optionally the steadystate density matrix
+    `rhoss` and a list of current superoperators `J_ops` can be specified. If
+    either of these are omitted they are computed internally.
+
+    Parameters
+    ----------
+
+    L : :class:`qutip.Qobj`
+        Qobj representing the system Liouvillian.
+
+    c_ops : array / list (optional)
+        List of current collapse operators.
+
+    rhoss : :class:`qutip.Qobj` (optional)
+        The steadystate density matrix corresponding the system Liouvillian
+        `L`.
+
+    J_ops : array / list (optional)
+        List of current superoperators.
+
+    Returns
+    --------
+    I : array
+        The currents `I` corresponding to each current collapse operator
+        `c_ops` (or, equivalently, each current superopeator `J_ops`).
+    """
+
+    if J_ops is None:
+        if c_ops is None:
+            raise ValueError("c_ops must be given if J_ops is not")
+        J_ops = [sprepost(c, c.dag()) for c in c_ops]
+
+    if rhoss is None:
+        if c_ops is None:
+            raise ValueError("c_ops must be given if rhoss is not")
+        rhoss = steadystate(L, c_ops)
+
+    rhoss_vec = _data.column_stack(rhoss.data.copy())
+
+    N = len(J_ops)
+    current = np.zeros(N)
+
+    for i, Ji in enumerate(J_ops):
+        current[i] = _data.expect_super(Ji.data, rhoss_vec).real
+
+    return current
+
+
+def _solve(A, V):
+    try:
+        if settings.has_mkl:
+            out = mkl_spsolve(A.tocsc(), V)
+        else:
+            A.sort_indices()
+            out = sp.linalg.splu(A, permc_spec='COLAMD').solve(V)
+    except Exception:
```

Comment:
This exception is raises when the sparse methods are used on very small matrices. I wonder if we should just do an explicit check on the matrix size? Trying the sparse method and failing seems a bit ugly.

@nwlambert Any thoughts on how / whether to clean this up since you wrote the original code?

Reply:
See 401076efb170 for @nwlambert's orignal commit message.

Reply:
I did a quick check, and yeah the old example i had in mind still fails sometimes without those exceptions to catch it.  It fails on a 2-dimensional system, so a Liouvillian with shape (4,4),  but is ok with 3-dimensions (shape (9,9)).  I guess other option is just to default to using lsqr() for the sparse case?

Reply:
Hmm. Computing the determinant of a large sparse matrix efficiently is hard, although there are some ideas mentioned here -- https://stackoverflow.com/questions/19107617/how-to-compute-scipy-sparse-matrix-determinant-without-turning-it-to-dense. I guess this is why `scipy.sparse.linalg.det` does not exist.

I'm okay with leaving this as is for now, but it would be good perhaps to leave the problem to the user by using a fast method and raising an error suggestion another method if the matrix is singular (and a way to specify that other method).

A similar problem arises in the HEOM steady_state function, so nice solution here could be used elsewhere in QuTiP too.

Reply:
`steadystate.py` is also full of multiple methods to solve this kind of problem. 
A `solve` method for `Qobj` could be nice...

Reply:
That's a good idea. Let's add it to our v5 todo list. Maybe even at the data layer?

Reply:
I added adding `solve` to the v5 alpha list.